[CA-939] validate method config TDR entity relationships

andy7i opened the pull request:
https://broadworkbench.atlassian.net/browse/CA-939

TDR now supports Relationships, which means that we can traverse tables in BQ. This PR provides the validation for relationships in MC expressions.

Note: `namespace: name` notation is not supported in BQ so this was removed.


- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [x] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

andy7i committed:
relationship traversal for workflow config

Code was partly implemented during https://github.com/broadinstitute/rawls/pull/1249

     // from DD: If the relationship exists than the tables must be valid. if they are not it is a bug!

// this is checking that the `fromTable` exists and that the relationship exists in one spot.  Not sure if we need a separate check for the table and another for the relationship
     // Maybe we don't need to do all that though...it feels like we're getting into the territory of validating the TDR
     // "relationship" schema at some point.  Maybe it's ok to just say, "If a relationship is defined with name X and
     // from_table FOO, then we're going to assume that the RelationshipModel object is valid for this SnapshotModel."

Co-Authored-By: Greg Polumbo <36171087+gpolumbo-broad@users.noreply.github.com>
Co-Authored-By: Marc Talbott <19961550+marctalbott@users.noreply.github.com>

andy7i committed:
added unit tests, errors ; todo: validateExpressionsForSubmission unit tests and figuring out why certain ones did not fail before

andy7i committed:
validateExpressionsForSubmission unit tests

andy7i committed:
refactor to disallow ':' in BigQuery table/column names

andy7i committed:
Revert "refactor to disallow ':' in BigQuery table/column names"

This reverts commit 2e839784

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/parser/antlr/DataRepoInputExpressionValidationVisitor.scala:
```diff
@@ -19,45 +22,72 @@ class DataRepoInputExpressionValidationVisitor(rootEntityType: Option[String],
     snapshotTables.find(_.getName == tableName)
   }
+
+  private def maybeGetNextTableFromRelation(fromTable: TableModel, relationName: String): Option[TableModel] = {
+    val relationships = snapshotModel.getRelationships.asScala.toList
+    maybeFindTableInSnapshotModel(fromTable.getName)
+      .flatMap { tableModel =>
+        relationships.find { relationship =>
+          relationship.getFrom.getTable == tableModel.getName && relationship.getFrom.getColumn == relationName
+        }.map { relationship =>
+          val nextTableName = relationship.getTo.getTable
+          return maybeFindTableInSnapshotModel(nextTableName)
+        }
+      }
+  }
```
3 things here:
1) try rewriting this as a for comprehension. More than one flapMap/map works better as a for comp
2) this needs to match the relationship in both directions. Either `relationship.getFrom.getTable == tableModel.getName` or `relationship.getTo.getTable == tableModel.getName`
3) `relationship.getFrom.getColumn == relationName` is not right, the relation name _is not_ the column name

andy7i committed:
overwrite visitNamespace for validation of namespace:name for BQ in DataRepoInputExpressionValidationVisitor

andy7i committed:
fixing tests

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/parser/antlr/DataRepoInputExpressionValidationVisitor.scala:
```diff
@@ -22,7 +25,32 @@ class DataRepoInputExpressionValidationVisitor(rootEntityType: Option[String],
+    }
+  }
+
+  override def visitNamespace(ctx: TerraExpressionParser.NamespaceContext): Try[Unit] = {
+    rootEntityType match {
+      case Some(_) => {
+        val maybeParent = Option(ctx.parent.parent.getChild(0).getText)
+        maybeParent match {
+          case Some(maybeWorkspace) => {
+            if (maybeWorkspace == "workspace.")
+              super.visitNamespace(ctx)
+            else
+              Failure(new RawlsExceptionWithErrorReport(
+                ErrorReport(StatusCodes.BadRequest, "Expressions with \"namespace: name\" are not valid for BigQuery")
+              ))
+          }
+          case None => Failure(new RawlsExceptionWithErrorReport(
+            ErrorReport(StatusCodes.BadRequest, "Expressions with \"namespace: name\" are not valid for BigQuery")
+          ))
+        }
+
+      }
+      case None => super.visitNamespace(ctx)
```
sorry, I gave you bad advice... I forgot we need to allow namespaces on workspace expressions.

The better place to do this is in `validateEntityLookup` where you can check `entityLookupContext.attributeName().namespace()`.

The code that looks like `ctx.parent.parent.getChild(0).getText` is the part that is concerning to me here. Also the checking for "workspace.". Both feel like brute force approaches that depend on the grammar structure but don't use the parseTree. If the grammar changes in the future these things may break and the compiler will miss it.

andy7i committed:
pr feedback -- in this.foo.bar, `foo` is a relationshipName not a column name.
  // todo: tests for multiple relationship traversal and from-to and to-from relationship

andy7i committed:
addressing feedback:
>this needs to match the relationship in both directions. Either relationship.getFrom.getTable == tableModel.getName or relationship.getTo.getTable == tableModel.getName

tests pass ; todo: more tests

andy7i committed:
check for namespace in validateEntityLookup. it seems like checking for null is not generally recommended in Scala but exists for backwards compatibility with Java. In this case, we are checking a Java object so it seems appropriate?

andy7i committed:
expand tests for forward and backward relationships

andy7i committed:
Merge branch 'develop' into aj--validate-relationship-mc--ca-939

# Conflicts:
#	project/Dependencies.scala

andy7i committed:
paired with Doug - changed options to Trys, optimized code

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/parser/antlr/DataRepoInputExpressionValidationVisitor.scala:
```diff
@@ -22,19 +24,23 @@ class DataRepoInputExpressionValidationVisitor(rootEntityType: Option[String],
-      case None => Failure(new RawlsException(s"Root entity type [$rootTableName] is not a name of a table that exist within DataRepo Snapshot."))
+      case (Failure(regrets), _) => Failure(regrets)
+      case (_, Some(_)) => Failure(new RawlsExceptionWithErrorReport(
+        ErrorReport(StatusCodes.BadRequest, "Expressions with \"namespace: name\" are not valid for BigQuery")
```
say "not valid with snapshots" in stead of "not valid for BigQuery"

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/parser/antlr/DataRepoInputExpressionValidationVisitor.scala:
```diff
@@ -44,20 +50,55 @@ class DataRepoInputExpressionValidationVisitor(rootEntityType: Option[String],
     } else {
-      Failure(new RawlsException(s"Missing attribute `${attributeName}` on table `${tableModel.getName}`"))
+      Failure(new RawlsExceptionWithErrorReport(
+        ErrorReport(StatusCodes.BadRequest, s"Missing attribute `${attributeName}` on table `${tableModel.getName}`")
```
error message should be `s"Column ${attributeName} does not exist on table ${tableModel.getName}"`

dvoet reviewed (approved):
a couple wording changes on error messages then 👍

andy7i committed:
PR feedback - error messages

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/parser/antlr/DataRepoInputExpressionValidationVisitor.scala:
```diff
@@ -22,19 +24,23 @@ class DataRepoInputExpressionValidationVisitor(rootEntityType: Option[String],
       case Some(rootTableName) => validateEntityLookup(rootTableName, ctx)
-      case None => Failure(new RawlsException("Expressions beginning with \"this.\" are only allowed when running with workspace data model. However, workspace attributes can be used."))
+      case None => Failure(new RawlsExceptionWithErrorReport(
+        ErrorReport(StatusCodes.BadRequest, "Expressions beginning with \"this.\" are only allowed when running with workspace data model. However, workspace attributes can be used.")
```
// it seems like it would get here only if methodConfiguration.rootEntityType is not set, so this error message doesn't make sense to me. Shouldn't it be something like "Missing Root Entity Type. Please select one and try again."

gpolumbo-broad replied:
I had to think about this too.  However, I think the point here is that here we're validating an individual Expression, not the MethodConfiguration.  And in particular, we know we're validating a DataRepo Expression, so I think this statement is ok.  Whether it could be wordsmithed for clarity is something maybe product or UX could help us understand if users know what the problem is and how to fix it when they see this message in context.

andy7i replied:
slack conversation about this: https://broadinstitute.slack.com/archives/GSHUGJ5RD/p1596826199127100

I'm okay with leaving it for now. It seemed off to me so I just wanted to make sure it was ok

andy7i committed:
self-review: co-locate relationship test setup ; fix clue

andy7i committed:
change some invalid cases to use real column names

andy7i committed:
Merge branch 'develop' into aj--validate-relationship-mc--ca-939

andy7i committed:
update SubmissionSpec test

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/parser/antlr/DataRepoInputExpressionValidationVisitor.scala:
```diff
@@ -22,19 +24,23 @@ class DataRepoInputExpressionValidationVisitor(rootEntityType: Option[String],
   private def validateEntityLookup(rootTableName: String, entityLookupContext: EntityLookupContext): Try[Unit] = {
-    maybeFindTableInSnapshotModel(rootTableName) match {
-      case Some(rootTableModel) => {
+    (maybeFindTableInSnapshotModel(rootTableName), Option(entityLookupContext.attributeName().namespace())) match {
```
I think this would read clearer as two separate checks.  First check if there's a namespace specified and if it's valid, then continue on to `maybeFindTableInSnapshotModel()`

I don't know if the FP/Scala purists agree with that approach, but I prefer the style of breaking down each step and calling it out explicitly.  I might even break out the namespace validation into its own method so I can name it, `validateEntityNamespace` to be explicitly clear about what that step is doing.

andy7i replied:
we actually had it as 2 separate checks at first, but opted for this approach to consolidate the various error states. This way it's more clear what the behavior is when it succeeds and when it fails. 

I do see your point about it being tougher to read! I think breaking out validateEntityNamespace would help make it easier to read and I'll give that a shot.

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/parser/antlr/DataRepoInputExpressionValidationVisitor.scala:
```diff
@@ -44,20 +50,55 @@ class DataRepoInputExpressionValidationVisitor(rootEntityType: Option[String],
-      case _ => Failure(new RawlsException(s"Relationships are not currently supported."))
+      case nextRelationContext :: remainingRelations => {
+        val nextRelationName = nextRelationContext.attributeName.name.getText
+        maybeGetNextTableFromRelation(currentTableModel, nextRelationName).flatMap(traverseRelationsAndGetFinalTable(_, remainingRelations))
```
I don't think this is tail recursive, unless scala does some magic with `flatMap` behind the scenes to make it so.

https://alvinalexander.com/scala/fp-book/tail-recursive-algorithms/

I don't think we need to worry about it here since most expressions I think would only result in a handful (literally single digits) of recursive calls.  That said, I would say it is good practice to always try to make your recursive functions tail recursive.  

You can try adding the `@tailrec` annotation to this method to see what the compiler thinks, maybe scala knows how to do something special with `flatMap` to make it tail recursive, but I don't know.  The way I'm reading this it is the flatten part of the `flatMap` that is going to cause the problem here.

andy7i replied:
yeah, it did have the tail recursion (that you added?) there before, but with the flatmap it gave an error:

```
Error:(62, 5) could not optimize @tailrec annotated method traverseRelationsAndGetFinalTable: it contains a recursive call not in tail position
    relations match {
```

Doug and I removed it since we figured (as you point out) that most expressions should be pretty short.

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityExpressionValidatorSpec.scala:
```diff
@@ -61,6 +62,30 @@ class DataRepoEntityExpressionValidatorSpec extends FlatSpec with TestDriverComp
+  val fromRootTableAndColumnToSecondTable: RelationshipTermModel = new RelationshipTermModel().table(rootTableName).column("root_second")
+  val fromSecondTableAndColumnToThirdTable: RelationshipTermModel = new RelationshipTermModel().table(secondTableName).column("second_third")
+  val fromSecondTableAndColumnToRootTable: RelationshipTermModel = new RelationshipTermModel().table(secondTableName).column("second_root")
+  val fromThirdTableAndColumnToSecondTable: RelationshipTermModel = new RelationshipTermModel().table(thirdTableName).column("third_second")
```
I'm having a hard time with these names and making sense out of what the schema looks like in my head.  Maybe describe the schema some other way in a comment?  And/or rename the columns?  Maybe consider some testing schema that we inherently already understand, like a couple of table of Books and Authors and Author Addresses or something like that, that the reader of this will inherently have some understanding of how it all ties together.

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityExpressionValidatorSpec.scala:
```diff
@@ -116,6 +141,33 @@ class DataRepoEntityExpressionValidatorSpec extends FlatSpec with TestDriverComp
+    val providerWithMultipleTables = createTestProvider(snapshotModel = createSnapshotModel(defaultFixtureTables, relationships))
+    val expressionValidatorWithMultipleTables: ExpressionValidator = providerWithMultipleTables.expressionValidator
+
+    val actualValid = expressionValidatorWithMultipleTables.validateMCExpressions(allValidWithRelationships, toGatherInputs(allValidWithRelationships.inputs)).futureValue
```
Naming this `actualValid` has tripped me a couple of times reading these tests.  Consider renaming to something like `validationResults`

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityExpressionValidatorSpec.scala:
```diff
@@ -116,6 +141,33 @@ class DataRepoEntityExpressionValidatorSpec extends FlatSpec with TestDriverComp
+    val expressionValidatorWithMultipleTables: ExpressionValidator = providerWithMultipleTables.expressionValidator
+
+    val actualValid = expressionValidatorWithMultipleTables.validateMCExpressions(allValidWithRelationships, toGatherInputs(allValidWithRelationships.inputs)).futureValue
+    actualValid.invalidInputs.size shouldBe 23
```
Same as above re: magic number

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/expressions/DataRepoExpressionFixture.scala:
```diff
@@ -34,22 +34,52 @@ trait DataRepoExpressionFixture {
     """["a",{"more":{"elaborate":this.gvcf}}]""",
     """{"more":{"elaborate":{"reference1": this.gvcf, "path":"gs://abc/123"}}}""",
-    """["foo", "bar", 123, ["array", this.gvcf, this.library:cohort], false]""",
-    """["foo", "bar", 123, ["array", this.gvcf, [this.library:cohort]], false, ["abc", this.with-dash]]"""
```
Why were these removed (and the others up on lines 37-39)?  I know that the namespaced attributes are not valid for DataRepo expressions, but these are valid expressions for the LocalEntityProvider.   This means we might need to creates some different lists that build for DataRepo vs. Local expressions.

andy7i replied:
I think they are separated. There's a Fixture for local expressions here: https://github.com/broadinstitute/rawls/blob/develop/core/src/test/scala/org/broadinstitute/dsde/rawls/expressions/LocalExpressionFixture.scala

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/expressions/DataRepoExpressionFixture.scala:
```diff
@@ -66,8 +96,17 @@ trait DataRepoExpressionFixture {
+    "this.rootTableToSecondTable.column_does_not_exist",
+    "this.library:cohort", // namespace:name is not allowed for BQ column/table names
+    "this.library:cohort1",
+    "this.arbitrary:whatever",
```
Again, these are only invalid TDR Entity Expressions

gpolumbo-broad replied:
I'm still concerned about the fact that we added these namespaced expressions to the list of invalid expressions because I think these same fixtures are used in both TDR and Local Entity Expression tests

marctalbott replied:
There's a different ExpressionFixture file called LocalExpressionFixture.scala so we can differentiate between what's valid for TDR and what's valid for the local entity provider

andy7i replied:
Maybe I'm missing something, but Intellij tells me that `DataRepoExpressionFixture` is only used by `DataRepoEntityExpressionValidatorSpec` and `LocalEntityLocalExpressionValidatorSpec` uses `LocalExpressionFixture`. Is there something else that I am not considering?

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/expressions/DataRepoExpressionFixture.scala:
```diff
@@ -116,10 +155,26 @@ trait DataRepoExpressionFixture {
+  val rootTableColumns = List("root_table_column", "gvcf", "with-dash", "underscores_are_ok", "_", "case_sample", "root_second")
+
+  val secondTableName = "secondTable"
+  val secondTableColumns = List("second_table_column", "second_third", "second_root")
```
If possible, would like to rename these columns again, as I struggle with this naming convention

andy7i committed:
pr feedback -- better naming of "actualValid" -> "validationResults"

andy7i committed:
pr feedback -- magic number

andy7i committed:
pr feedback -- human relatable tables/relationships

andy7i committed:
add diagram for relationship tables, move to DataRepoExpressionFixture

andy7i committed:
PR feedback - refactor validateEntityLookup to use 2 checks (with a new validateEntityNamespace) instead of 1 combined match

andy7i committed:
comments

gpolumbo-broad reviewed (commented):
Maybe we need to get together with @marctalbott too and talk about how to sort out the valid/invalid expressions in the fixtures?

andy7i committed:
Merge branch 'develop' into aj--validate-relationship-mc--ca-939

# Conflicts:
#	project/Dependencies.scala

andy7i commented:
jenkins retest -- `bucket did not become readable`
